### PR TITLE
Fix SFT training for prompt-completion type and transformers v5

### PR DESCRIPTION
**Pull request by qgallouedec**
## What this PR does?

This PR fix a bug due to the change of default value of `return_dict` in `transformers` library. See https://github.com/huggingface/transformers/pull/41626 Before version 5.0.0, the default value of `return_dict` is `False`, while in version 5.0.0 and later, it is `True`. This causes an almost silent bug in `SFTTrainer` when .

This PR also sorts the arguments of `apply_chat_template` method to make it consistent with the order of arguments the method.

## What changed in transformers?

```python
>>> from transformers import AutoTokenizer
>>> tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen3-0.6B")
>>> messages = [
...     {"role": "user", "content": "What is the meaning of life?"},
...     {"role": "assistant", "content": "The final answer is..."},
... ]
>>> tokenizer.apply_chat_template(messages, tokenize=True, return_dict=False)
[151644, 872, 198, 3838, 374, 279, 7290, 315, 2272, 30, 151645, 198, 151644, 77091, 198, 151667, 271, 151668, 271, 785, 1590, 4226, 374, 1112, 151645, 198]
>>> tokenizer.apply_chat_template(messages, tokenize=True, return_dict=True)
{'input_ids': [151644, 872, 198, 3838, 374, 279, 7290, 315, 2272, 30, 151645, 198, 151644, 77091, 198, 151667, 271, 151668, 271, 785, 1590, 4226, 374, 1112, 151645, 198], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}
```

### transformers<5.0.0

```python
>>> tokenizer.apply_chat_template(messages, tokenize=True)
[151644, 872, 198, 3838, 374, 279, 7290, 315, 2272, 30, 151645, 198, 151644, 77091, 198, 151667, 271, 151668, 271, 785, 1590, 4226, 374, 1112, 151645, 198]
```

### transformers>=5.0.0

```python
>>> tokenizer.apply_chat_template(messages, tokenize=True)
{'input_ids': [151644, 872, 198, 3838, 374, 279, 7290, 315, 2272, 30, 151645, 198, 151644, 77091, 198, 151667, 271, 151668, 271, 785, 1590, 4226, 374, 1112, 151645, 198], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}
```

Note that this is only an issue when `tokenize=True` (default). When `tokenize=False`, the output format is consistent across `transformers` versions.

To keep backward compatibility, you should explicitly set the value of `return_dict` when calling `apply_chat_template`.

## To reproduce

install transformers v5

```python
from trl import SFTTrainer, SFTConfig
from datasets import Dataset

dataset = Dataset.from_dict(
    {
        "prompt": [
            [{"role": "user", "content": "This is my prompt."}],
            [{"role": "user", "content": "This is my second prompt."}],
        ],
        "completion": [
            [{"role": "assistant", "content": "This is my response."}],
            [{"role": "assistant", "content": "This is my second response."}],
        ],
    }
)

trainer = SFTTrainer(model="Qwen/Qwen3-0.6B", train_dataset=dataset)
```

You'll get

```
Mismatch between tokenized prompt and the start of tokenized prompt+completion. This may be due to unexpected tokenizer behavior, whitespace issues, or special token handling. Verify that the tokenizer is processing text consistently.
Mismatch between tokenized prompt and the start of tokenized prompt+completion. This may be due to unexpected tokenizer behavior, whitespace issues, or special token handling. Verify that the tokenizer is processing text consistently.
```

but the training won't fail. But if you check the dataset:

```python
>>> trainer.train_dataset["completion_mask"]
Column([[0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], [0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])
```

only the two first tokens are masked, which is not correct. After the fix:

```python
>>> trainer.train_dataset["completion_mask"]
Column([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])
```

why 2? because the number of tokens to mask is computed using `len(prompt_ids)`

https://github.com/huggingface/trl/blob/96021dee9e38ab005a774c2cbfcc33c91e506388/trl/trainer/sft_trainer.py#L1052

but when `return_dict=True`, `prompt_ids` is a dict with 2 keys.

### CODE REVIEW

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -1001,20 +1001,21 @@ def tokenize_fn(example, processing_class, dataset_text_field, assistant_only_lo
                                 completion = example["completion"]
                             prompt_ids = processing_class.apply_chat_template(
                                 prompt,
-                                tokenize=True,
-                                add_generation_prompt=True,
                                 tools=example.get("tools"),
+                                add_generation_prompt=True,
+                                tokenize=True,
+                                return_dict=False,
```
Comment (qgallouedec) [inline trl/trainer/sft_trainer.py:1007]: This is the only critical line change of this PR, the rest is QOL

**Review by albertvillanova — APPROVED** [review]
Good catch! Thanks.

**Comment by afarahani-sharethis** [comment]
@qgallouedec thanks for this fix. 
I don't see #4880 fix is released on 0.27.0. 
As far as I understood this can cause a major issue if a user is finetuning for completion. 
I had to patch this manually on my side after installing 0.27.0.
Any plan for releasing this officially?

**Comment by qgallouedec** [comment]
Yep I'll do a patch release today
